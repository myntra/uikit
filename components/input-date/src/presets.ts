import dayJS from 'dayjs'
import { onlyDate } from './input-date-utils'

export const TODAY = {
  range: false,
  label: 'Today',
  value: () => onlyDate(new Date()),
}

export const YESTERDAY = {
  range: false,
  label: 'Yesterday',
  value: () => onlyDate(dayJS().subtract(1, 'day')),
}

export const CURRENT_WEEK = {
  range: true,
  label: 'Current Week',
  value: () => ({
    from: onlyDate(dayJS().startOf('week')), // TODO(bug): Selects wrong date range.
    to: onlyDate(dayJS()),
  }),
}

export const LAST_7_DAYS = {
  range: true,
  label: 'Last 7 Days',
  value: () => ({
    from: onlyDate(dayJS().subtract(6, 'day')),
    to: onlyDate(dayJS()),
  }),
}

export const LAST_30_DAYS = {
  range: true,
  label: 'Last 30 Days',
  value: () => ({
    from: onlyDate(dayJS().subtract(29, 'day')),
    to: onlyDate(dayJS()),
  }),
}

export const LAST_WEEK = {
  range: true,
  label: 'Last Week',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'week')
        .startOf('week')
        .day(1)
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'week')
        .endOf('week')
    ),
  }),
}

export const LAST_MONTH = {
  range: true,
  label: 'Last Month',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'month')
        .startOf('month')
        .add(1, 'day')
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'month')
        .endOf('month')
    ),
  }),
}

export const LAST_YEAR = {
  range: true,
  label: 'Last Year',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'year')
        .startOf('year')
        .add(1, 'day')
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'year')
        .endOf('year')
    ),
  }),
}
