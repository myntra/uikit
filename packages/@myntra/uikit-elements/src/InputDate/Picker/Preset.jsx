import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { isDateEqual } from '../InputDateUtils'
import classnames from './Preset.module.css'

/**
 A preset date selection list.

 @since 0.0.0
 @status REVIEWING
 @example
 <InputDatePicker.Preset presets={[{ label: 'Today', value: () => new Date() }]} />
 */
class Preset extends PureComponent {
  static propTypes = {
    presets: PropTypes.arrayOf(
      PropTypes.shape({
        range: PropTypes.bool,
        label: PropTypes.string.isRequired,
        /**
         * @function
         * @returns {Date|{ from: Date, to: Date }}
         */
        value: PropTypes.func.isRequired
      })
    ).isRequired,
    value: PropTypes.oneOfType([
      PropTypes.instanceOf(Date),
      PropTypes.shape({
        from: PropTypes.instanceOf(Date),
        to: PropTypes.instanceOf(Date)
      })
    ]),
    range: PropTypes.bool,
    onChange: PropTypes.func.isRequired
  }

  handleChange = preset => {
    return this.props.onChange(this.normalize(preset.value()))
  }

  normalize(value) {
    if (this.props.range && value instanceof Date) {
      return { from: value, to: value }
    }

    return value
  }

  isActive(preset) {
    if (this.props.range) {
      const value = this.normalize(preset.value())

      return (
        this.props.value && isDateEqual(this.props.value.from, value.from) && isDateEqual(this.props.value.to, value.to)
      )
    }

    return isDateEqual(this.props.value, preset.value())
  }

  render() {
    return (
      <div className={classnames('presets')}>
        {this.props.presets
          .filter(preset => preset.range === undefined || preset.range === this.props.range)
          .map(preset => (
            <div
              role="button"
              tabIndex={-1}
              key={preset.label}
              onClick={() => this.handleChange(preset)}
              className={classnames('preset', { active: this.isActive(preset) })}
            >
              {preset.label}
            </div>
          ))}
      </div>
    )
  }
}

export default Preset
