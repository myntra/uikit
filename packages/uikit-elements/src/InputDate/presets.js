import dayJS from 'dayjs'
import { onlyDate } from './InputDateUtils'

export const TODAY = {
  label: 'Today',
  value: () => onlyDate(new Date())
}

export const YESTERDAY = {
  label: 'Yesterday',
  value: () => onlyDate(dayJS().subtract(1, 'day'))
}

export const CURRENT_WEEK = {
  range: true,
  label: 'Current Week',
  value: () => ({
    from: onlyDate(dayJS().startOf('week')), // TODO(bug): Selects wrong date range.
    to: onlyDate(dayJS())
  })
}

export const LAST_7_DAYS = {
  range: true,
  label: 'Last 7 Days',
  value: () => ({
    from: onlyDate(dayJS().subtract(7, 'day')),
    to: onlyDate(dayJS().subtract(1, 'day'))
  })
}

export const LAST_30_DAYS = {
  range: true,
  label: 'Last 30 Days',
  value: () => ({
    from: onlyDate(dayJS().subtract(30, 'day')),
    to: onlyDate(dayJS().subtract(1, 'day'))
  })
}

export const LAST_WEEK = {
  range: true,
  label: 'Last Week',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'week')
        .startOf('week')
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'week')
        .endOf('week')
    )
  })
}

export const LAST_MONTH = {
  range: true,
  label: 'Last Month',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'month')
        .startOf('month')
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'month')
        .endOf('month')
    )
  })
}

export const LAST_YEAR = {
  range: true,
  label: 'Last Year',
  value: () => ({
    from: onlyDate(
      dayJS()
        .subtract(1, 'year')
        .startOf('year')
    ),
    to: onlyDate(
      dayJS()
        .subtract(1, 'year')
        .endOf('year')
    )
  })
}
