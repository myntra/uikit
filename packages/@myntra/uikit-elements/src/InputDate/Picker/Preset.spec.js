import React from 'react'
import { shallow } from 'enzyme'
import Preset from './Preset'
import { YESTERDAY, LAST_7_DAYS } from '../presets'
import dayJS from 'dayjs'

const yesterday = YESTERDAY.value()

describe('render', () => {
  it('should render', () => {
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} onChange={jest.fn()} />)

    expect(wrapper.find('.preset')).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('should highlight preset', () => {
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} value={yesterday} onChange={jest.fn()} />)

    expect(wrapper.find('.active')).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render range', () => {
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} onChange={jest.fn()} range />)

    expect(wrapper.find('.preset')).toHaveLength(2)
    expect(wrapper).toMatchSnapshot()
  })

  it('should highlight preset range', () => {
    const yesterday = dayJS()
      .subtract(1, 'day')
      .toDate()
    const wrapper = shallow(
      <Preset
        presets={[YESTERDAY, LAST_7_DAYS]}
        value={{ from: yesterday, to: yesterday }}
        range
        onChange={jest.fn()}
      />
    )

    expect(wrapper.find('.active')).toHaveLength(1)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('action', () => {
  it('should select', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} onChange={fn} />)

    wrapper
      .find('.preset')
      .at(0)
      .simulate('click')

    expect(fn).toHaveBeenCalledWith(yesterday)
  })

  it('should select (single)', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} onChange={fn} range />)

    wrapper
      .find('.preset')
      .at(0)
      .simulate('click')

    expect(fn).toHaveBeenCalledWith({ from: yesterday, to: yesterday })
  })

  it('should render (range)', () => {
    const fn = jest.fn()
    const wrapper = shallow(<Preset presets={[YESTERDAY, LAST_7_DAYS]} onChange={fn} range />)

    wrapper
      .find('.preset')
      .at(1)
      .simulate('click')

    expect(fn).toHaveBeenCalledWith(LAST_7_DAYS.value())
  })
})
