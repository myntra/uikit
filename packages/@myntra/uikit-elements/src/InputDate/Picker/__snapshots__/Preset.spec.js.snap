// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should highlight preset 1`] = `
<div
  className="presets"
>
  <div
    className="preset active"
    key="Yesterday"
    onClick={[Function]}
  >
    Yesterday
  </div>
</div>
`;

exports[`render should highlight preset range 1`] = `
<div
  className="presets"
>
  <div
    className="preset active"
    key="Yesterday"
    onClick={[Function]}
  >
    Yesterday
  </div>
  <div
    className="preset"
    key="Last 7 Days"
    onClick={[Function]}
  >
    Last 7 Days
  </div>
</div>
`;

exports[`render should render 1`] = `
<div
  className="presets"
>
  <div
    className="preset"
    key="Yesterday"
    onClick={[Function]}
  >
    Yesterday
  </div>
</div>
`;

exports[`render should render range 1`] = `
<div
  className="presets"
>
  <div
    className="preset"
    key="Yesterday"
    onClick={[Function]}
  >
    Yesterday
  </div>
  <div
    className="preset"
    key="Last 7 Days"
    onClick={[Function]}
  >
    Last 7 Days
  </div>
</div>
`;
